src/emulator/storage/apis/firebase.ts

```diff
@@ -352,8 +353,9 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
         }
         throw err;
       }
-      metadata.addDownloadToken();
-      return res.status(200).json(new OutgoingFirebaseMetadata(metadata));
+
+      storedMetadata.addDownloadToken();
```

Comment:
Just peeked at our server implementation; interestingly enough, we only generate the download token on metadata get, not object get (data fetch). The metadata get handler actually updates GCS with the new download token, so I have a hunch the server will actually trigger any metadata update functions that are set up.

The rest of this PR looks good to me but can you verify the behavior of the server and mimic it here? I just want to make sure we limit the amount of divergence between the emulator and the server.

Reply:
Yup that makes sense thanks for the callout. Here's all the places I was able to find that we add download tokens:

- Metadata Get ([Link](https://source.corp.google.com/piper///depot/google3/java/com/google/firebase/storage/requesthandlers/MetadataGetHandler.java;l=47;rcl=436357765))
  - We generate a new download token if one isn't present when we fetch metadata
- Object Creation ([Link](https://source.corp.google.com/piper///depot/google3/java/com/google/firebase/storage/requesthandlers/ObjectCreateHandler.java;l=200;bpv=1;bpt=1;rcl=436357765))
- If we delete a download token and it was the last one, we replace it with a fresh token ([Link](https://source.corp.google.com/piper///depot/google3/java/com/google/firebase/storage/requesthandlers/TokenDeleteHandler.java;l=52;rcl=436357765))